### PR TITLE
Fix bug with DB-Forge composite foreign keys

### LINKED ISSUE #4310: Bug: DB Forge does not support adding composite foreign keys
**Direction**
It's impossible to add composite foreign key using forge

**Describe the bug**
I need to add the foreign key on multiple columns like this

```
  CONSTRAINT fk_column
     FOREIGN KEY (column1, column2)
     REFERENCES parent_table (column1, column2)
     ON UPDATE CASCADE
     ON DELETE CASCADE
```
Executing forge produces error when following is used

```
$this->forge->addForeignKey('column1,column2', 'parent_table', 'column1,column2', 'cascade', 'cascade')
```

**CodeIgniter 4 version**
dev-master

**Affected module(s)**
Database

This Issue is just to be noted. I'll work on the solution when I'm available. 

Related to:
#4120 
#4115

### PR DESCRIPTION
**Pull request by monkenWu**
**Description**
Fixes #4310 
Hi, I edited the `addForeignKey` method and `_processForeignKeys` method of `CodeIgniter\Database\Forge` so that they can define composite foreign keys in an array.

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by samsonasik** [comment]
Please run cs fixer

```bash
vendor/bin/php-cs-fixer fix system/Database/Forge.php
vendor/bin/php-cs-fixer fix tests/system/Database/Live/ForgeTest.php
```

**Comment by monkenWu** [comment]
@samsonasik It's done.

**Comment by samsonasik** [comment]
Please run rector as well:

```bash
vendor/bin/rector
```

**Comment by monkenWu** [comment]
@samsonasik 
Hi, I fixed the problem of `_processForeignKeys` method when using `SQLSRV` Driver, modified the test case, and ran `vendor/bin/rector process` .

File: system/Database/Forge.php [hunk 1]
```diff
@@ -403,27 +403,40 @@ public function addField($field)
...
+
+        if (is_array($fieldName)) {
+            foreach ($fieldName as $name) {
+                if (! isset($this->fields[$name])) {
+                    $errorNames[] = $name;
+                }
+            }
+        } elseif (! isset($this->fields[$fieldName])) {
+            $errorNames[] = $fieldName;
         }
```
Comment (paulbalandan) [inline system/Database/Forge.php:428]: You can just simply force this to array then iterate over. No need for if else condition.
```php
$fieldName = (array) $fieldName;

foreach ($fieldName as $name) {
    if (! isset($this->fields[$name])) {
        $errorNames[] = $name;
    }
}
```

File: system/Database/Forge.php [hunk 2]
```diff
@@ -403,27 +403,40 @@ public function addField($field)
...
 
-        $this->foreignKeys[$fieldName] = [
-            'table'    => $tableName,
-            'field'    => $tableField,
-            'onDelete' => strtoupper($onDelete),
-            'onUpdate' => strtoupper($onUpdate),
+        if (! empty($errorNames)) {
```
Comment (paulbalandan) [inline system/Database/Forge.php:430]:
```suggestion
        if ($errorNames !== []) {
```

File: system/Database/Forge.php [hunk 3]
```diff
@@ -1208,11 +1221,25 @@ protected function _processForeignKeys(string $table): string
...
+                if (is_array($fkey['field'])) {
+                    $nameIndex = $table . '_' . implode('_', $fkey['field']) . '_foreign';
+                } else {
+                    $nameIndex = $table . '_' . $fkey['field'] . '_foreign';
+                }
+                if (is_array($fkey['field'])) {
+                    $foreignKeyFiled = implode(', ', $this->db->escapeIdentifiers($fkey['field']));
+                } else {
+                    $foreignKeyFiled = $this->db->escapeIdentifiers($fkey['field']);
+                }
```
Comment (paulbalandan) [inline system/Database/Forge.php:1234]: Since `field` is now an array, you can dispense the if else.

File: system/Database/Forge.php [hunk 4]
```diff
@@ -1208,11 +1221,25 @@ protected function _processForeignKeys(string $table): string
...
+                if (is_array($fkey['referenceField'])) {
+                    $referenceField = implode(', ', $this->db->escapeIdentifiers($fkey['referenceField']));
+                } else {
+                    $referenceField = $this->db->escapeIdentifiers($fkey['referenceField']);
+                }
```
Comment (paulbalandan) [inline system/Database/Forge.php:1239]: You can possibly force convert to array to dispense the if else.

File: system/Database/Forge.php [hunk 5]
```diff
@@ -1208,11 +1221,25 @@ protected function _processForeignKeys(string $table): string
...
 
                 $sql .= ",\n\tCONSTRAINT " . $this->db->escapeIdentifiers($nameIndex)
-                    . ' FOREIGN KEY(' . $this->db->escapeIdentifiers($field) . ') REFERENCES ' . $this->db->escapeIdentifiers($this->db->DBPrefix . $fkey['table']) . ' (' . $this->db->escapeIdentifiers($fkey['field']) . ')';
+                    . ' FOREIGN KEY(' . $foreignKeyFiled . ') REFERENCES ' . $this->db->escapeIdentifiers($this->db->DBPrefix . $fkey['referenceTable']) . ' (' . $referenceField . ')';
```
Comment (paulbalandan) [inline system/Database/Forge.php:1242]: Can you use `sprintf` here to easily see the format of the SQL you're creating?

File: system/Database/SQLSRV/Forge.php [hunk 6]
```diff
@@ -233,46 +233,6 @@ protected function _processColumn(array $field): string
...
     }
 
-    /**
-     * Process foreign keys
-     *
-     * @param string $table Table name
-     *
-     * @return string
-     */
-    protected function _processForeignKeys(string $table): string
```
Comment (paulbalandan) [inline system/Database/SQLSRV/Forge.php:243]: You cannot simply delete a protected method as userland code might be depending on this. If this is no longer needed, then deprecate the method.

**Review by paulbalandan — COMMENTED** [review]
Irrespective of the position whether the change in signature will be accepted or not, I have left some comments. And please change the base branch to `codeigniter4:4.2` instead of `develop`.

File: system/Database/Forge.php [hunk 7]
```diff
@@ -1208,11 +1221,25 @@ protected function _processForeignKeys(string $table): string
...
-            foreach ($this->foreignKeys as $field => $fkey) {
-                $nameIndex = $table . '_' . $field . '_foreign';
+            foreach ($this->foreignKeys as $fkey) {
+                if (is_array($fkey['field'])) {
+                    $nameIndex = $table . '_' . implode('_', $fkey['field']) . '_foreign';
+                } else {
+                    $nameIndex = $table . '_' . $fkey['field'] . '_foreign';
+                }
+                if (is_array($fkey['field'])) {
+                    $foreignKeyFiled = implode(', ', $this->db->escapeIdentifiers($fkey['field']));
```
Comment (samsonasik) [inline system/Database/Forge.php:1231]: I think "Filed" should be "Filled" with double `l`

**Comment by monkenWu** [comment]
@paulbalandan  @samsonasik

Thanks for the review.
I corrected some spelling errors, improved code quality, and rolledback the `_processIndexes` method of the `SQLSRV` class.

**Comment by paulbalandan** [comment]
@monkenWu it seems your PR branch's tail is still based on `develop`. Please rebase this against the current `4.2` branch so that changes only from your PR will appear here. If anything went wrong,, you can just recreate your PR against `4.2`

**Comment by monkenWu** [comment]
Change the branch base to 4.2.

**Pull request closed without merging**